rclcpp/src/rclcpp/clock.cpp

```diff
@@ -182,6 +182,64 @@ Clock::sleep_for(Duration rel_time, Context::SharedPtr context)
   return sleep_until(now() + rel_time, context);
 }
 
+bool
+Clock::started()
+{
+  if (!rcl_clock_valid(get_clock_handle())) {
+    throw std::runtime_error("clock is not rcl_clock_valid");
+  }
+  return rcl_clock_time_started(get_clock_handle());
+}
+
+bool
+Clock::wait_until_started(Context::SharedPtr context)
+{
+  if (!context || !context->is_valid()) {
+    throw std::runtime_error("context cannot be slept with because it's invalid");
+  }
+  if (!rcl_clock_valid(get_clock_handle())) {
+    throw std::runtime_error("clock cannot be waited on as it is not rcl_clock_valid");
+  }
+
+  if (started()) {
+    return true;
+  } else {
+    // Wait until the first non-zero time
+    return sleep_until(rclcpp::Time(0, 1, get_clock_type()), context);
+  }
+}
+
+bool
+Clock::wait_until_started(
+  const Duration & timeout,
+  Context::SharedPtr context,
+  const Duration & wait_tick_ns)
+{
+  if (!context || !context->is_valid()) {
+    throw std::runtime_error("context cannot be slept with because it's invalid");
+  }
+  if (!rcl_clock_valid(get_clock_handle())) {
+    throw std::runtime_error("clock cannot be waited on as it is not rcl_clock_valid");
+  }
+
+  Clock timeout_clock = Clock(RCL_STEADY_TIME);
+  Time start = timeout_clock.now();
+
+  while (!started() && context->is_valid()) {  // Context check checks for rclcpp::shutdown()
+    if (timeout < wait_tick_ns) {
+      timeout_clock.sleep_for(timeout);
+    } else {
+      timeout_clock.sleep_for(Duration(wait_tick_ns));
+    }
+
+    if (timeout_clock.now() - start > timeout) {
+      return started();
+    }
+  }
```

Comment:
if `timeout` is 91s, `wait_tick_ns` is 10s, last sleep will be 10s that is 9 seconds off from the user expectation. i think it needs to calculate how much time left to wait for this case.

Reply:
@methylDragon can you check this comment?

Reply:
Ah, right! I need to get to this, oops

Reply:
https://github.com/ros2/rclcpp/pull/2040/commits/1df3675219280b460527de2cf3564b903f321c24 !